File: packages/flask/korea_auth/result.py
"""
Result class for authentication operations
"""

from dataclasses import dataclass
from typing import Optional, Dict, Any


@dataclass
class AuthResult:
    """Result of an authentication verification attempt."""
    
    success: bool
    uid: Optional[str] = None
    provider: Optional[str] = None
    name: Optional[str] = None
    email: Optional[str] = None
    photo_url: Optional[str] = None
    raw_data: Optional[Dict[str, Any]] = None
    error: Optional[str] = None
    
    @classmethod
    def success_result(
        cls,
        uid: str,
        provider: str,
        name: Optional[str] = None,
        email: Optional[str] = None,
        photo_url: Optional[str] = None,
        raw_data: Optional[Dict[str, Any]] = None,
    ) -> "AuthResult":
        """Create a successful result."""
        return cls(
            success=True,
            uid=uid,
            provider=provider,
            name=name,
            email=email,
            photo_url=photo_url,
            raw_data=raw_data,
        )
    
    @classmethod
    def failure(cls, error: str) -> "AuthResult":
        """Create a failed result."""
        return cls(success=False, error=error)

File: packages/rails/lib/korea_auth/result.rb
# frozen_string_literal: true

module KoreaAuth
  class Result
    attr_reader :uid, :provider, :name, :email, :photo_url, :raw_data, :error

    def initialize(success:, uid: nil, provider: nil, name: nil, email: nil, photo_url: nil, raw_data: nil, error: nil)
      @success = success
      @uid = uid
      @provider = provider
      @name = name
      @email = email
      @photo_url = photo_url
      @raw_data = raw_data
      @error = error
    end

    def success?
      @success
    end

    def failure?
      !@success
    end

    def self.success(uid:, provider:, name: nil, email: nil, photo_url: nil, raw_data: nil)
      new(success: true, uid: uid, provider: provider, name: name, email: email, photo_url: photo_url, raw_data: raw_data)
    end

    def self.failure(error)
      new(success: false, error: error)
    end
  end
end
